File: Problem66_NumberOfIslands.java
// Time Complexity : O(m * n)
// Space Complexity : O(m * n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    boolean[][] visited;
    int m;
    int n;
    public int numIslands(char[][] grid) {
        this.m = grid.length;
        this.n = grid[0].length;
        visited = new boolean[m][n];
        int islandCount = 0;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i][j] == '1' && !visited[i][j]) {
                    islandCount++;
                    dfs(grid, i, j);
                }
            }
        }
        return islandCount;
    }
    public void dfs(char[][] grid, int row, int col) {
        if(row >= 0 && row < m && col >= 0 && col < n && !visited[row][col] && grid[row][col] == '1') {
            visited[row][col] = true;
            dfs(grid, row + 1, col);
            dfs(grid, row - 1, col);
            dfs(grid, row, col + 1);
            dfs(grid, row, col - 1);
        }
    }
}

File: Problem-66 Number of Islands.py
# 200. Number of Islands
# https://leetcode.com/problems/number-of-islands/

# Logic: Iterate the matrix. When you find a 1, start BFS and increment the island count.

# Time Complexiety: O(n*m)
# Space Complexiety: O(min(n,m))

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])
        islandCount = 0
        q = list()
        dirs = [(0,1),(0,-1),(1,0),(-1,0)]
        
        for i in range(n):
            for j in range(m):
                if grid[i][j] == '1':
                    q.append((i,j))
                    grid[i][j] = '2'
                    islandCount += 1
                    
                    while q:
                        row, col = q.pop(0)
                        
                        for _dir in dirs:
                            newRow = row + _dir[0]
                            newCol = col + _dir[1]
                            
                            if 0 <= newRow < n and 0 <= newCol < m:
                                if grid[newRow][newCol] == "1":
                                    q.append((newRow,newCol))
                                    grid[newRow][newCol] = "2"
        return islandCount